src/lower/mod.rs

```diff
@@ -1018,29 +1018,46 @@ impl Anonymize for [ir::ParameterKind<ir::Identifier>] {
     }
 }
 
-macro_rules! well_formed {
-    ($self: ident, $value: expr, $constructor: expr, $program: expr) => {{
-        let where_clauses = $self.binders.value.where_clauses
-                  .iter()
-                  .cloned()
-                  .flat_map(|wc| wc.expanded($program));
+trait ExpandWhereClauses {
+    fn expanded_where_clauses(&self, program: &ir::Program) -> Vec<ir::DomainGoal>;
+}
 
-        let where_clauses: Vec<ir::DomainGoal> = where_clauses.collect();
+impl ExpandWhereClauses for ir::StructDatumBound {
+    fn expanded_where_clauses(&self, program: &ir::Program) -> Vec<ir::DomainGoal> {
+        self.where_clauses
+            .iter()
+            .cloned()
+            .flat_map(|wc| wc.expanded(program))
+            .collect()
+    }
+}
+
+impl ExpandWhereClauses for ir::TraitDatumBound {
+    fn expanded_where_clauses(&self, program: &ir::Program) -> Vec<ir::DomainGoal> {
+        self.where_clauses
+            .iter()
+            .cloned()
+            .flat_map(|wc| wc.expanded(program))
+            .collect()
+    }
+}
+
+trait WellFormed {
+    fn wf_clauses(&self, wf_predicate: ir::WellFormed, program: &ir::Program)
+        -> Vec<ir::ProgramClause>;
+}
+
+impl<T: ExpandWhereClauses> WellFormed for ir::Binders<T> {
+    fn wf_clauses(&self, wf_predicate: ir::WellFormed, program: &ir::Program)
+        -> Vec<ir::ProgramClause>
+    {
+        let where_clauses = self.value.expanded_where_clauses(program);
 
         let wf = ir::ProgramClause {
-            implication: $self.binders.map_ref(|_| {
+            implication: self.map_ref(|_| {
                 ir::ProgramClauseImplication {
-                    consequence: $constructor($value.clone().cast()).cast(),
-                    conditions: {
-                        let tys = $value.parameters
-                                        .iter()
-                                        .filter_map(|pk| pk.as_ref().ty())
-                                        .cloned()
-                                        .map(|ty| ir::WellFormed::Ty(ty))
-                                        .casted();
-
-                        tys.chain(where_clauses.iter().cloned().casted()).collect()
-                    }
+                    consequence: wf_predicate.clone().cast(),
+                    conditions: where_clauses.iter().cloned().casted().collect(),
```

Comment:
in particular it's not obvious to me what has changed in this commit other than tests, just looking at the code anyhow